docs/question/server.md

```diff
@@ -1,89 +1,312 @@
+
+It is recommended that you give additional feedback to the student as they make progress towards the solution, and reward this progress with partial credit.
+
+If this function is not defined, the question will be graded automatically based on the correct answers set in `data["correct_answers"]`. Each answer the student provides will also be given feedback from the element that graded it. If the `grade` function _is_ defined, the data you receive has already been graded by the elements. You should ensure you only award partial credit if the answer isn't correct, otherwise you might give partial credit for a correct answer. In the snipped below, we update `data["score"]` using the [`set_weighted_score_data`][prairielearn.question_utils.set_weighted_score_data] utility.
 
-<!-- y is defined by data["correct_answers"]["y"] in server.py's `generate()`. -->
+You can set `data["format_errors"][NAME]` to mark the submission as invalid. This will cause the question to not use up one of the student's attempts on assessments. However, you are encouraged to do as many checks for invalid data as possible in `parse` instead of `grade`; the `parse` function is called when the student hits "Save only", in manually graded questions, and in assessments without real-time grading.
+
+```python title="server.py"
+import math
+import prairielearn as pl
+
+def grade(data):
+    # Give half points for incorrect answers larger than "x", only if not already correct.
+    # Use math.isclose to avoid possible floating point errors.
+    y_is_correct = math.isclose(data["partial_scores"]["y"]["score"], 1.0)
+    if not y_is_correct and data["submitted_answers"]["y"] > data["params"]["x"]:
+        data["partial_scores"]["y"]["score"] = 0.5
+        pl.set_weighted_score_data(data)
+        data["feedback"]["y"] = "Your value for $y$ is larger than $x$, but incorrect."
+```
+
+### Providing feedback
+
+To set custom feedback, the grading function should set the corresponding entry in the `data["feedback"]` dictionary. These feedback entries are passed in when rendering the `question.html`, which can be accessed by using the mustache prefix `{{feedback.}}`. See the [above example](#complete-example) or [this demo question](https://github.com/PrairieLearn/PrairieLearn/tree/master/exampleCourse/questions/demo/custom/gradeFunction) for examples of this.
+
+Some elements provide feedback through `data["partial_scores"][NAME]["feedback"]`, which you can override in the grading function. This field is a string, except for the `<pl-drawing>` and `<pl-checkbox>` elements. The feedback provided here will be attached to the student's answer for that element, which can make it easier for students to interpret the feedback they received in longer questions. Since not elements provide feedback this way, and use different data structures to represent the feedback they give, we recommend using `data["feedback"][NAME]` to provide question-specific feedback instead.
```

Comment:
```suggestion
Some elements provide feedback through `data["partial_scores"][NAME]["feedback"]`, which you can override in the grading function. This field is often a string, though some elements such as `<pl-drawing>` and `<pl-checkbox>` use different types. The feedback provided here will be attached to the student's answer for that element, which can make it easier for students to interpret the feedback they received in longer questions. Since not elements provide feedback this way, and use different data structures to represent the feedback they give, we recommend using `data["feedback"][NAME]` to provide question-specific feedback instead.
```

Reply:
This is just future proofing? I would prefer to give definites.

Reply:
It's partly future proof, but also to cover course-level custom elements.

Reply:
I'd prefer to phrase it in a way that ensures the instructor checks what's appropriate instead of relying on this to update the feedback blindly.

Reply:
How about a qualifier like `Some [official] elements`, so that it's clear we aren't talking about course-level custom elements?

Reply:
I'd still prefer to have it clear that there is no implied consistency and that question writers should confirm before assuming.

Reply:
I think this ties into what we are discussing on the slack. about the standardization of the `partial_scores` interface.

Reply:
I'm fine making this change, but it would be nice to talk about this in #dev-core a little more first.

Reply:
Since there wasn't much discussion, I'll add the change you suggested, and maybe we can revisit this down the road.

Reply:
I agree with @jonatanschroeder 's change here. 

Additionaly, with regards to this part:

> we recommend using `data["feedback"][NAME]` to provide question-specific feedback instead.

do we specify somewhere that they should insert that into the HTML themselves if they are going to do that?

Reply:
We do. https://prairielearn--11830.org.readthedocs.build/en/11830/question/server/#providing-feedback
![image](https://github.com/user-attachments/assets/55960e9e-8ef5-419f-afee-da3285f228bb)